src/jailer/src/env.rs

```diff
@@ -168,20 +171,34 @@ impl Env {
 
         SyscallReturnCode(unsafe { libc::chown(dev_path.as_ptr(), self.uid(), self.gid()) })
             .into_empty_result()
-            .map_err(|e| Error::ChangeFileOwner(dev_path.to_str().unwrap(), e))
+            // Safe to unwrap as we provided valid file names.
+            .map_err(|e| Error::ChangeFileOwner(PathBuf::from(dev_path.to_str().unwrap()), e))
     }
 
-    fn setup_jailed_folders(&mut self) -> Result<()> {
-        for folder in FOLDER_HIERARCHY.iter() {
-            let folder_cstr = CStr::from_bytes_with_nul(folder).map_err(Error::FromBytesWithNul)?;
-
-            // This unwrap is safe as we provided strings that have valid utf8 chars.
+    fn setup_jailed_folders<P>(&self, folders: Vec<P>) -> Result<()>
+    where
+        P: AsRef<Path> + AsRef<OsStr>,
+    {
+        for folder in folders.iter() {
+            let mut folder_bytes = Path::new(folder).as_os_str().as_bytes().to_vec();
+            folder_bytes.push(b'\0');
```

Comment:
nit: 
Using `&str` and dynamically pushing the 'NULL-terminating character' is more extensible, but it is also slower code on the hotpath..

Seeing how we're always setting up static things (like chroot hierarchy in this function), I would optimize the jailer code for speed rather than flexibility.

Reply:
+1

Also this complicates the logic a bit.

Reply:
I'll try to figure out a way to specify the dir names with the nul terminator and without copying them around. My purpose with this change was to make `setup_jailed_folders` reusable (for the far future of an independent `jailer` crate) and unit test-able (for the immediate future of this PR).

What's the hot path you're referring to? This is done once per microvm, so I'm guessing you're thinking of the process startup time? That hasn't been exceeded (we have an integ test) but even so I'm not thrilled with the extra string operations so I'll try to get rid of them anyway.

Reply:
> I'm guessing you're thinking of the process startup time? That hasn't been exceeded (we have an integ test)

You are right, that's why I marked this as a `nit`.

I am thinking in the context of future-proof work. The process start-up time we have right now is not great to begin with, and I think there's work to do to improve it. In that context, I would avoid changes that are (even theoretically) slower.

Reply:
@acatangiu @serban300 I reworked that part to get rid of extra string copying and processing, PTAL.